_posts/2020-06-10-connect-kafka-spring-kafka.md

```diff
@@ -0,0 +1,507 @@
+We can also retrieve the useful metadata about the consumed record using `@Header()` annotation as shown in the above example.
+
+### Adding Filters to Listeners
+
+Spring provides a strategy to filter messages before reaching to our listeners.
+
+```java
+class KafkaConsumerConfig {
+
+    @Bean
+    KafkaListenerContainerFactory<ConcurrentMessageListenerContainer<String, String>>
+    kafkaListenerContainerFactory() {
+        ConcurrentKafkaListenerContainerFactory<String, String> factory = new ConcurrentKafkaListenerContainerFactory<>();
+        factory.setConsumerFactory(consumerFactory());
+
+        factory.setRecordFilterStrategy(record -> record.value().contains("ignored"));
+        return factory;
+    }
+}
+```
+
+Spring wraps the listener with `FilteringMessageListenerAdapter`. It takes an implementation of `RecordFilterStrategy` in which we implement the filter method and the records which won't match get discarded before reaching the listener.
+
+In the above example, we have added a filter to discard the records which contain the word "ignored".
+
+### Reply using `@SendTo`
+
+Spring allows to send method's return value to specified destination with `@SendTo`.
+
+```java
+@KafkaListener(topics = "reflectoring-others")
+@SendTo("reflectoring-1")
+String listenAndReply(String message) {
+
+  LOG.info("ListenAndReply [{}]", message);
+  return "This is a reply sent after receiving message";
+}
+```
+
+SpringBoot default factory configuration gives the SendTo template. Since we are overriding the factory configuration in our code example, the listener container factory must be provided with a `KafkaTemplate` by using `setReplyTemplate` which is used to send the reply. In the above example, we are forwarding the reply message to the topic "reflectoring-1".
+
+## Custom Records
+
+Let's now look at how to send/receive a Java Object as a JSON byte[].
+
+```java
+class User {
+    private String name;
+  ...
+}
+```
+
+We'll be sending and receiving the above `User` class in our example.
+
+### Using JSON Serializer & Deserializer
+
+To achieve this, we must configuring our producer and consumer to use JSON serializer & deserializer.
+
+```java
+@Configuration
+class KafkaProducerConfig {
+    ...
+
+    @Bean
+    public ProducerFactory<String, User> userProducerFactory() {
+        ...
+        configProps.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, JsonSerializer.class);
+        return new DefaultKafkaProducerFactory<>(configProps);
+    }
+
+    @Bean
+    public KafkaTemplate<String, User> userKafkaTemplate() {
+        return new KafkaTemplate<>(userProducerFactory());
+    }
+}
+```
+
+```java
+@Configuration
+class KafkaConsumerConfig {
+    ...
+    public ConsumerFactory<String, User> userConsumerFactory() {
+        ...
+        return new DefaultKafkaConsumerFactory<>(props, new StringDeserializer(), new JsonDeserializer<>(User.class));
```

Comment:
Actually I have given ellipses since similar example was given already in Consumer Configuration section. 

For better perspective I removed ellipses and added code.